ext/node/zlib/mod.rs

```diff
@@ -0,0 +1,474 @@
+  if condition {
+    Ok(())
+  } else {
+    Err(type_error(msg.to_string()))
+  }
+}
+
+struct ZlibInner {
+  dictionary: Option<Vec<u8>>,
+  err: i32,
+  flush: i32,
+  init_done: bool,
+  level: i32,
+  mem_level: i32,
+  mode: i32,
+  strategy: i32,
+  window_bits: i32,
+  write_in_progress: bool,
+  pending_close: bool,
+  gzib_id_bytes_read: u32,
+  strm: z_stream,
+}
+
+const GZIP_HEADER_ID1: u8 = 0x1f;
+const GZIP_HEADER_ID2: u8 = 0x8b;
+
+impl ZlibInner {
+  fn start_write(
+    &mut self,
+    input: &[u8],
+    in_off: u32,
+    in_len: u32,
+    out: &mut [u8],
+    out_off: u32,
+    out_len: u32,
+    flush: i32,
+  ) -> Result<(), AnyError> {
+    check(self.init_done, "write before init")?;
+    check(self.mode != NONE, "already finialized")?;
+    check(!self.write_in_progress, "write already in progress")?;
+    check(!self.pending_close, "close already in progress")?;
+
+    self.write_in_progress = true;
+
+    if flush != Z_NO_FLUSH
+      && flush != Z_PARTIAL_FLUSH
+      && flush != Z_SYNC_FLUSH
+      && flush != Z_FULL_FLUSH
+      && flush != Z_FINISH
+      && flush != Z_BLOCK
+    {
+      return Err(type_error("Bad argument"));
+    }
+
+    self.strm.avail_in = in_len;
+    // TODO(@littledivy): Crap! Guard against overflow!!
+    self.strm.next_in = unsafe { input.as_ptr().offset(in_off as isize) } as _;
+    self.strm.avail_out = out_len;
+    self.strm.next_out =
+      unsafe { out.as_mut_ptr().offset(out_off as isize) } as _;
+
+    self.flush = flush;
+    Ok(())
+  }
+
+  fn do_write(&mut self, flush: i32) -> Result<(), AnyError> {
+    match self.mode {
+      DEFLATE | GZIP | DEFLATERAW => {
+        self.err = unsafe { libz_sys::deflate(&mut self.strm, flush) };
+      }
+      UNZIP if self.strm.avail_in > 0 => {
+        let mut next_expected_header_byte = Some(0);
+
+        if self.gzib_id_bytes_read == 0 {
+          let byte = unsafe { *self.strm.next_in.offset(0) };
+          if byte == GZIP_HEADER_ID1 {
+            self.gzib_id_bytes_read = 1;
+            next_expected_header_byte = Some(1);
+
+            // Not enough.
+            if self.strm.avail_in == 1 {
+              self.write_in_progress = false;
+              return Ok(());
+            }
+          } else {
+            self.mode = INFLATE;
+            next_expected_header_byte = None;
+          }
+        }
+
+        if self.gzib_id_bytes_read == 1 && next_expected_header_byte.is_some() {
+          let byte = unsafe {
+            *self.strm.next_in.offset(next_expected_header_byte.unwrap())
+          };
+          if byte == GZIP_HEADER_ID2 {
+            self.gzib_id_bytes_read = 2;
+            self.mode = GUNZIP;
+          } else {
+            self.mode = INFLATE;
+          }
+        } else if next_expected_header_byte != None {
+          return Err(type_error(
+            "invalid number of gzip magic number bytes read",
+          ));
+        }
+      }
+      _ => {}
```

Comment:
Nevermind - I think see what's going on in the code. I wonder if it would be clearer to handle `UNZIP` first, and then have a `match` on all modes? It's a bit confusing with the change of modes.